`dotnet pack` mistakenly uses alias for WPF Library projects

When packing a project with a target platform, when the target platform version is not included in the .csproj (but a platform is), the lib files are pulled in using the "alias".

## Details about Problem

NuGet product used (NuGet.exe | VS UI | Package Manager Console | dotnet.exe): All of the above

dotnet.exe --version (if appropriate): 5.0.100-rc.1.20454.5

VS version (if appropriate): 16.8.4-pre.4.0

OS version (i.e. win10 v1607 (14393.321)): win10

Worked before? If so, with which NuGet version: no

## Detailed repro steps so we can see the same problem

1. mkdir nuget-repro && cd nuget-repro && dotnet new wpflib
2. dotnet pack
3. look in the .nupkg lib directory

## Result

We get a facefull of warnings:

```
❯ dotnet pack
Microsoft (R) Build Engine version 16.8.0-preview-20451-02+51a1071f8 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  All projects are up-to-date for restore.
  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
  bar -> C:\Users\kmarchan\src\bar\bin\Debug\net5.0-windows\bar.dll
C:\Program Files\dotnet\sdk\5.0.100-rc.1.20454.5\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(198,5): warning NU5128: Some target frameworks declared in the dependencies group of the nuspec and the lib/ref folder do not have exact matches in the other location. Consult the list of actions below: [C:\Users\kmarchan\src\bar\bar.csproj]
C:\Program Files\dotnet\sdk\5.0.100-rc.1.20454.5\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(198,5): warning NU5128: - Add a dependency group for net5.0-windows to the nuspec [C:\Users\kmarchan\src\bar\bar.csproj]
C:\Program Files\dotnet\sdk\5.0.100-rc.1.20454.5\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(198,5): warning NU5128: - Add lib or ref assemblies for the net5.0-windows7.0 target framework [C:\Users\kmarchan\src\bar\bar.csproj]
  Successfully created package 'C:\Users\kmarchan\src\bar\bin\Debug\bar.1.0.0.nupkg'.
```

...and it's using `net5.0-windows` instead of `net5.0-windows7.0`.

--- a/src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTaskLogic.cs
+++ b/src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTaskLogic.cs
@@ -55,6 +55,12 @@ namespace NuGet.Build.Tasks.Pack
                 packArgs.MinClientVersion = version;
             }
 
+            LockFile assetsFile = GetAssetsFile(request);
+            var aliases = new Dictionary<string, string>();
+            foreach (var tfm in assetsFile.PackageSpec.TargetFrameworks)
+            {
+                aliases[tfm.TargetAlias] = tfm.FrameworkName.GetShortFolderName();
+            }
 
             InitCurrentDirectoryAndFileName(request, packArgs);
             InitNuspecOutputPath(request, packArgs);
@@ -76,12 +82,12 @@ namespace NuGet.Build.Tasks.Pack
                 // This only needs to happen when packing via csproj, not nuspec.
                 packArgs.PackTargetArgs.AllowedOutputExtensionsInPackageBuildOutputFolder = InitOutputExtensions(request.AllowedOutputExtensionsInPackageBuildOutputFolder);
                 packArgs.PackTargetArgs.AllowedOutputExtensionsInSymbolsPackageBuildOutputFolder = InitOutputExtensions(request.AllowedOutputExtensionsInSymbolsPackageBuildOutputFolder);
-                packArgs.PackTargetArgs.TargetPathsToAssemblies = InitLibFiles(request.BuildOutputInPackage);
-                packArgs.PackTargetArgs.TargetPathsToSymbols = InitLibFiles(request.TargetPathsToSymbols);
+                packArgs.PackTargetArgs.TargetPathsToAssemblies = InitLibFiles(request.BuildOutputInPackage, aliases);
+                packArgs.PackTargetArgs.TargetPathsToSymbols = InitLibFiles(request.TargetPathsToSymbols, aliases);
                 packArgs.PackTargetArgs.AssemblyName = request.AssemblyName;
                 packArgs.PackTargetArgs.IncludeBuildOutput = request.IncludeBuildOutput;
                 packArgs.PackTargetArgs.BuildOutputFolder = request.BuildOutputFolders;
-                packArgs.PackTargetArgs.TargetFrameworks = ParseFrameworks(request);
+                packArgs.PackTargetArgs.TargetFrameworks = ParseFrameworks(request, aliases);
 
                 if (request.IncludeSource)
                 {
@@ -322,6 +328,28 @@ namespace NuGet.Build.Tasks.Pack
             return version;
         }
 
+
+        private LockFile GetAssetsFile(IPackTaskRequest<IMSBuildItem> request)
+        {
+            if (request.PackItem == null)
+            {
+                throw new PackagingException(NuGetLogCode.NU5028, Strings.NoPackItemProvided);
+            }
+
+            string assetsFilePath = Path.Combine(request.RestoreOutputPath, LockFileFormat.AssetsFileName);
+
+            if (!File.Exists(assetsFilePath))
+            {
+                throw new InvalidOperationException(string.Format(
+                    CultureInfo.CurrentCulture,
+                    Strings.AssetsFileNotFound,
+                    assetsFilePath));
+            }
+            // The assets file is necessary for project and package references. Pack should not do any traversal,
+            // so we leave that work up to restore (which produces the assets file).
+            var lockFileFormat = new LockFileFormat();
+            return lockFileFormat.Read(assetsFilePath);
+        }
         private void PopulateFrameworkAssemblyReferences(PackageBuilder builder, IPackTaskRequest<IMSBuildItem> request)
         {
             // First add all the assembly references which are not specific to a certain TFM.
@@ -396,7 +424,7 @@ namespace NuGet.Build.Tasks.Pack
             return runner.RunPackageBuild();
         }
 
-        private IEnumerable<OutputLibFile> InitLibFiles(IMSBuildItem[] libFiles)
+        private IEnumerable<OutputLibFile> InitLibFiles(IMSBuildItem[] libFiles, IDictionary<string, string> aliases)
         {
             var assemblies = new List<OutputLibFile>();
             if (libFiles == null)
@@ -410,7 +438,7 @@ namespace NuGet.Build.Tasks.Pack
                 var finalOutputPath = assembly.GetProperty("FinalOutputPath");
 
                 // Fallback to using Identity if FinalOutputPath is not set.
-                // See bug https://github.com/NuGet/Home/issues/5408 
+                // See bug https://github.com/NuGet/Home/issues/5408
                 if (string.IsNullOrEmpty(finalOutputPath))
                 {
                     finalOutputPath = assembly.GetProperty(IdentityProperty);
@@ -422,6 +450,13 @@ namespace NuGet.Build.Tasks.Pack
                 if (!File.Exists(finalOutputPath))
                 {
                     throw new PackagingException(NuGetLogCode.NU5026, string.Format(CultureInfo.CurrentCulture, Strings.Error_FileNotFound, finalOutputPath));
+                }
+
+                string translated = null;
+                var succeeded = aliases.TryGetValue(targetFramework, out translated);
+                if (succeeded)
+                {
+                    targetFramework = translated;
                 }
 
                 // If target path is not set, default it to the file name. Only satellite DLLs have a special target path
@@ -448,12 +483,20 @@ namespace NuGet.Build.Tasks.Pack
             return assemblies;
         }
 
-        private ISet<NuGetFramework> ParseFrameworks(IPackTaskRequest<IMSBuildItem> request)
+        private ISet<NuGetFramework> ParseFrameworks(IPackTaskRequest<IMSBuildItem> request, IDictionary<string, string> aliases)
         {
             var nugetFrameworks = new HashSet<NuGetFramework>();
             if (request.TargetFrameworks != null)
             {
-                nugetFrameworks = new HashSet<NuGetFramework>(request.TargetFrameworks.Select(t => NuGetFramework.Parse(t)));
+                nugetFrameworks = new HashSet<NuGetFramework>(request.TargetFrameworks.Select(targetFramework => {
+                    string translated = null;
+                    var succeeded = aliases.TryGetValue(targetFramework, out translated);
+                    if (succeeded)
+                    {
+                        targetFramework = translated;
+                    }
+                    return NuGetFramework.Parse(targetFramework);
+                }));
             }
 
             return nugetFrameworks;
@@ -655,7 +698,7 @@ namespace NuGet.Build.Tasks.Pack
                     var newTargetPath = Path.Combine(targetPath, identity);
                     // We need to do this because evaluated identity in the above line of code can be an empty string
                     // in the case when the original identity string was the absolute path to a file in project directory, and is in
-                    // the same directory as the csproj file. 
+                    // the same directory as the csproj file.
                     newTargetPath = PathUtility.EnsureTrailingSlash(newTargetPath);
                     newTargetPaths.Add(newTargetPath);
                 }
@@ -813,7 +856,7 @@ namespace NuGet.Build.Tasks.Pack
 
                     var versionToUse = new VersionRange(targetLibrary.Version);
 
-                    // Use the project reference version obtained at build time if it exists, otherwise fallback to the one in assets file. 
+                    // Use the project reference version obtained at build time if it exists, otherwise fallback to the one in assets file.
                     if (projectRefToVersionMap.TryGetValue(projectReference.ProjectPath, out var projectRefVersion))
                     {
                         versionToUse = VersionRange.Parse(projectRefVersion, allowFloating: false);
@@ -872,7 +915,7 @@ namespace NuGet.Build.Tasks.Pack
                 // Add each package dependency.
                 foreach (var packageDependency in packageDependencies)
                 {
-                    // If we have a floating package dependency like 1.2.3-xyz-*, we 
+                    // If we have a floating package dependency like 1.2.3-xyz-*, we
                     // use the version of the package that restore resolved it to.
                     if (packageDependency.LibraryRange.VersionRange.IsFloating)
                     {

--- a/test/NuGet.Core.Tests/NuGet.Build.Tasks.Pack.Test/PackTaskLogicTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.Build.Tasks.Pack.Test/PackTaskLogicTests.cs
@@ -500,6 +500,121 @@ namespace NuGet.Build.Tasks.Pack.Test
             }
         }
 
+        [Fact]
+        public void PackTaskLogic_InfersFrameworkPlatformVersionFromAlias()
+        {
+            // Arrange
+            using (var testDir = TestDirectory.Create())
+            {
+                var tc = new TestContext(testDir, "net5.0-windows");
+
+                var assetsJson = @"{
+                    ""version"": 3,
+  ""targets"": {
+    ""net5.0"": {},
+    ""net5.0-windows7.0"": {}
+  },
+  ""libraries"": {},
+  ""projectFileDependencyGroups"": {
+    ""net5.0"": [],
+    ""net5.0-windows7.0"": []
+  },
+  ""project"": {
+    ""version"": ""0.0.0"",
+    ""restore"": {
+      ""projectName"": ""bar"",
+      ""projectStyle"": ""PackageReference"",
+      ""crossTargeting"": true,
+      ""fallbackFolders"": [
+        ""C:\\Microsoft\\Xamarin\\NuGet\\""
+      ],
+      ""originalTargetFrameworks"": [
+        ""net5.0"",
+        ""net5.0-windows""
+      ],
+      ""sources"": {
+        ""https://api.nuget.org/v3/index.json"": {},
+      },
+      ""frameworks"": {
+        ""net5.0"": {
+          ""targetAlias"": ""net5.0"",
+          ""projectReferences"": {}
+        },
+        ""net5.0-windows7.0"": {
+          ""targetAlias"": ""net5.0-windows"",
+          ""projectReferences"": {}
+        }
+      },
+      ""warningProperties"": {
+        ""warnAsError"": [
+          ""NU1605""
+        ]
+      }
+    },
+    ""frameworks"": {
+      ""net5.0"": {
+        ""targetAlias"": ""net5.0"",
+        ""imports"": [
+          ""net461"",
+          ""net462"",
+          ""net47"",
+          ""net471"",
+          ""net472"",
+          ""net48""
+        ],
+        ""assetTargetFallback"": true,
+        ""warn"": true,
+        ""frameworkReferences"": {
+          ""Microsoft.NETCore.App"": {
+            ""privateAssets"": ""all""
+          }
+        },
+      },
+      ""net5.0-windows7.0"": {
+        ""targetAlias"": ""net5.0-windows"",
+        ""imports"": [
+          ""net461"",
+          ""net462"",
+          ""net47"",
+          ""net471"",
+          ""net472"",
+          ""net48""
+        ],
+        ""assetTargetFallback"": true,
+        ""warn"": true,
+        ""frameworkReferences"": {
+          ""Microsoft.NETCore.App"": {
+            ""privateAssets"": ""all""
+          }
+        },
+      }
+    }
+  }
+                }";
+                File.WriteAllText(Path.Combine(testDir, "obj", "project.assets.json"), assetsJson);
+
+                // var msbuildItem = tc.AddContentToProject("", "abc.txt", "hello world");
+                // tc.Request.PackageFiles = new MSBuildItem[] { msbuildItem };
+                // tc.Request.ContentTargetFolders = new string[] { "content" };
+
+                // Act
+                tc.BuildPackage();
+
+                // Assert
+                using (var nupkgReader = new PackageArchiveReader(tc.NupkgPath))
+                {
+                    var nuspecReader = nupkgReader.NuspecReader;
+
+                    // Validate the assets.
+                    var libItems = nupkgReader.GetLibItems().ToList();
+                    Assert.Equal(1, libItems.Count);
+                    Assert.Equal(NuGetFramework.Parse("net5.0-windows7.0"), libItems[0].TargetFramework);
+                    Assert.Equal(new[] { "lib/net5.0-windows7.0/a.dll" }, libItems[0].Items);
+                }
+            }
+        }
+
+
         [PlatformTheory(Platform.Windows)]
         [InlineData(true)]
         [InlineData(false)]
@@ -625,11 +740,18 @@ namespace NuGet.Build.Tasks.Pack.Test
 
         private class TestContext
         {
+
             public TestContext(TestDirectory testDir)
+                : this(testDir, "net45")
+                {
+
+                }
+
+            public TestContext(TestDirectory testDir, string tfm)
             {
                 var fullPath = Path.Combine(testDir, "project.csproj");
                 var rootDir = Path.GetPathRoot(testDir);
-                var dllDir = Path.Combine(testDir, "bin", "Debug", "net45");
+                var dllDir = Path.Combine(testDir, "bin", "Debug", tfm);
                 var dllPath = Path.Combine(dllDir, "a.dll");
 
                 Directory.CreateDirectory(dllDir);
@@ -664,11 +786,11 @@ namespace NuGet.Build.Tasks.Pack.Test
                     IncludeBuildOutput = true,
                     RestoreOutputPath = Path.Combine(testDir, "obj"),
                     ContinuePackingAfterGeneratingNuspec = true,
-                    TargetFrameworks = new[] { "net45" },
+                    TargetFrameworks = new[] { tfm },
                     BuildOutputInPackage = new[] { new MSBuildItem(dllPath, new Dictionary<string, string>
                     {
                         {"FinalOutputPath", dllPath },
-                        {"TargetFramework", "net45" }
+                        {"TargetFramework", tfm }
                     })},
                     Logger = new TestLogger(),
                     SymbolPackageFormat = "symbols.nupkg",

